Ability to reposition and modify ui elements in more detail

YawLighthouse opened the pull request:
Hello I've been working on the loading screen plugin for my own game and have been adding to it, and thought it would be good to share it with everybody! So Im submitting this pull request for a ton of features that allow for more control over this plugin.

- Ability to re-position ui elements (its based off a canvas being the root essentially so all you need to do is make a Widget Blueprint and place widgets on the canvas like you normally would and just copy their slot values into the loading screen settings)
- Ability to toggle visibility on each widget object ahead of time(cause not everybody would want tips on a specific kind of loading screen)
- Ability to modify the text's font, size, color and opacity, AND MORE! :D
- Ability to have ui elements play during or after movies play(this also works the same way with images)
- Separated aspects of each element into their own structures on the C++ side to make it easier to modify and add to and 
- Ability to have a delay between tips before it automatically updates to a new random one(if you just want one to show up per loading screen I already took care of that functionality for you ;) )

I think thats about it I spent a good chunk of time basically reworking it to be more modifiable, let me know what you think if you or if you found any problems with it, I just wanna make this plugin as useful as possible. :)

(Sorry if the commit list for this is a little vague, was doing minor changes that I forgot to add until the last minute)

YawLighthouse committed:
Updated

Updated

Commit:
Merge pull request #1 from Oldsiren/Update

Updated

YawLighthouse committed:
Update

Update

YawLighthouse committed:
Update Cleanup

Cleaned up

lucasdlk commented:
Your update make my UE 4.22.1 project crash at startup :(

Oldsiren commented:
@lucasdlk ok, I have 2 things to ask, is it possible I could get an error from you (so I can find out where it crashed and fix it)? And I tested it on the launcher version of 4.22.2 and the source version, maybe try updating and see what happens?

Commit:
Fixed startup crash

Fixed crash where slate application was not valid on startup, replaced it with FPlatformTime.

Oldsiren commented:
@lucasdlk K try now, if it crashes again please give me the error code :)

Commit:
Added const correctness

Constant for a boolean that isnt getting changed anywhere else

Zod20 commented:
@Oldsiren works for me and its awesome! Can you tell me how to do 'wait for manual stop'? I can't seem to find a blueprint node to call from within my code to stop it.

Oldsiren commented:
@Zod20 Manual stop only gets used when Stop Movie is called(which because its during a loading screen its not really possible to run any blueprint code during that time) and when a any key is pressed(which is used by the MoviesAreSkippable boolean and only works if its actually finished loading and if that boolean is true) if you want a movie to stop playing when it finishes loading, then set AutoCompleteWhenLoadingCompletes to true. Honestly I dont think Wait for manual stop is actually useful for this plugin or even used but I kept it in because maybe in the future it will be.

lucasdlk commented:
Ok, i've updated my project to 4.22.2, still crash at startup. Here is the error :

> `Assertion failed: IsValid() [File:c:\program files\epic games\ue_4.22\engine\source\runtime\core\public\Templates/SharedPointer.h] [Line: 797]
UE4Editor_LoadingScreen!SSimpleLoadingScreen::Construct() [f:\unrealprojects\altexp\altexp_ue4\plugins\loadingscreen-master\source\loadingscreen\private\ssimpleloadingscreen.cpp:166]
UE4Editor_LoadingScreen!FLoadingScreenModule::BeginLoadingScreen() [f:\unrealprojects\altexp\altexp_ue4\plugins\loadingscreen-master\source\loadingscreen\private\loadingscreenmodule.cpp:122]
UE4Editor_LoadingScreen!FLoadingScreenModule::StartupModule() [f:\unrealprojects\altexp\altexp_ue4\plugins\loadingscreen-master\source\loadingscreen\private\loadingscreenmodule.cpp:65]`

Here is a link to full error log :
https://pastebin.com/tDGSRbUS

Commit:
Fixed crash from tooltips

Fixed crash from tooltips not having any to display. Where the widget was not created when it should have been

Oldsiren commented:
@lucasdlk ok I found the bug, it was related to the tooltips. It should be good to go now, let me know if it works :)

lucasdlk commented:
It seems that your update work. But now, i can't find the "Loading Screen" tab on Project Settings. Is it on purpose ?

Commit:
Added image support for throbber

An image(or material) can be used in place of a throbber

lucasdlk commented:
Can you make an option to randomize tips ?

Oldsiren commented:
@lucasdlk it does randomize tips automatically, if you want it to randomize it while loading the mouse has to be moving in the app(currently trying to figure out how to fix it so it works regardless of mouse movement)

lucasdlk commented:
At startup, tips at index 0 always showing first

professionaljones commented:
Hey there, I'm trying to set the position of the text  but I'm having a hard time getting it in the right place. Is there any way to see the widget locations without having to start a standalone process?

Oldsiren commented:
@professionaljones at the moment my solution is to just make an empty widget blueprint and plan out the placement there with a canvas panel being the root(because thats what Im using in this branch of the plugin) and then just copy those values over into the plugin's settings

Byrkoet commented:
Settings tips and changing `MinimumLoadingScreenDisplayTime` from `-1.000` to anything above 0 causes the widget to hand and only hide/collapse when input is detected :/

Oldsiren commented:
@Byrkoet so thats something with the ue4 loading screen system, basically this plugin talks with stuff thats already in the engine that handles the loading screen and one of those in engine settings is the `MinimumLoadingScreenDisplayTime` variable. Unless @nickdarnell would be able to provide a better answer or solution to what ever your trying to do, I'm not going to be of much help... (btw Nick thank you for the plugin, from one Nick to another ;) )

DW-Sebastien commented:
@Oldsiren this looks very promising, thanks for sharing!

I will try to update to your version in the process of upgrading our project to Unreal Engine 4.22 (should happen in the coming days)

In the meantime, I would suggest you rename this Pull Request with a clear title, stating what it does to all the people looking at this :)

YawLighthouse committed:
Fixed Include paths to account for wherever module is

Just an ease of use thing that fixes any warning or issues where it cannot find the public/private include paths when compiling from source.

YawLighthouse committed:
Forgot to include the proper libraries

Fixed system.io library not being included, formatted file to be with the way build files for unreal are currently formatted and setup

SIBarbu commented:
Hi! First of all I must say that this plugin, and then this branch are incredibly awesome. I had my own loading screen WBP until now, and this is just heaven because you don't have to do anything. Although, I have 2 questions:
1. For some reason, the center justification doesn't work at all on my text. Could it be an issue on my end?
2. I know this is not the intended use, but sometimes you might want to show loading screens before level travels. (Creating lobbies, loading data, etc.) Do you think there could be a way to show this loading screen at will with either C++ or Blueprints? I would just keep my current loading screen WBP, but the transition between them is not that pleasant.
Thank you!

YawLighthouse committed:
Add function library and added the ability to get and set the loading screen settings

YawLighthouse committed:
Updated to 4.23

Let me know of any bugs :)

Commit:
Merge pull request #2 from Oldsiren/4.23

Updated to 4.23

Commit:
Update README.md

Commit:
4.24 Update

Commit:
Remove engine specific line

avfxproton commented:
Hi Oldsiren,

I have two feature requests, and then I think your branch would be perfect.

1, Ability to fade in and fade out the loading screen on begin and end.
2, Ability to trigger the loading screen via blueprint not only when opening levels, but when doing some more complex level streaming, and then end the loading screen via blueprint aswell.
3, The third is not really a feature request, but rather a question. The startup loading screen takes 3-4 seconds even on a blank startup level. Is there some artificial timer for this? And is there a way how to turn off this behavior?

I would take a swing at adding these features myself, but then I would not want to merge my changes with each new updates, and if u decided to do it anyway, I think you would code it much differently than I would.

Thanks a lot for your great work!
